config/src/main/java/io/scalecube/config/JsonDocumentConfigPropertyImpl.java

```diff
@@ -0,0 +1,80 @@
+package io.scalecube.config;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectReader;
+import io.scalecube.config.source.LoadedConfigProperty;
+import io.scalecube.config.utils.ObjectMapperHolder;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link ObjectConfigProperty}.
+ *
+ * @param <T> type of the property value
+ */
+class JsonDocumentConfigPropertyImpl<T> extends AbstractConfigProperty<T>
+    implements ObjectConfigProperty<T> {
+
+  protected volatile T value;
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(JsonDocumentConfigPropertyImpl.class);
+  private final ObjectReader reader;
+
+  JsonDocumentConfigPropertyImpl(
+      String documentKey,
+      Class<T> cfgClass,
+      Map<String, LoadedConfigProperty> propertyMap,
+      Map<String, Map<Class, PropertyCallback>> propertyCallbackMap) {
+
+    super(documentKey, cfgClass);
+    reader = ObjectMapperHolder.getInstance().readerFor(cfgClass);
+    PropertyCallback<T> propertyCallback = new PropertyCallback<>(this::valueParser);
+
+    setPropertyCallback(propertyCallback);
+    synchronized (propertyCallbackMap) {
+      propertyCallbackMap
+          .computeIfAbsent(documentKey, k -> new ConcurrentHashMap<>())
+          .put(cfgClass, propertyCallback);
+    }
+    computeValue(
+        Collections.singletonList(
+            propertyMap.getOrDefault(
+                documentKey, LoadedConfigProperty.forNameAndValue(documentKey, null))));
+  }
+
+  private T valueParser(List<LoadedConfigProperty> properties) {
+    return properties
+        .stream()
+        .filter(this::isMyProperty)
+        .findFirst()
+        .flatMap(LoadedConfigProperty::valueAsString)
+        .map(this::parse)
+        .orElse(null);
+  }
+
+  private T parse(String source) {
```

Comment:
Replace this method with:

```
  private T parse(String source) {
    try {
      if (source != null) {
        return reader.readValue(source);
      }
    } catch (Exception ignoredException) {
      throw new RuntimeException(ignoredException);
    }
    return null;
  }
```

we shouldn't decide that problem with the old value, just throw an exception like it does Integer parser when we put no number value.

Reply:
The configuration data should be _sterile_ so I assume that the exceptions are **`IOExceptions`** rather than **`JSONParseException`**

Reply:
Ok, you can come up with some custom exception which will be extended `RuntimeException`.